src/main/java/uk/gov/justice/digital/delius/service/UserService.java

```diff
@@ -119,6 +119,38 @@ public UserDetailsWrapper getUserDetailsList(final Set<String> usernames) {
                         .build());
     }
 
+    public List<UserDetails> getUserDetailsByEmail(final String email) {
+        // first we perform the LDAP search to get a list of matching records for the email address
+        final var userList = ldapRepository.getDeliusUserByEmail(email);
+
+        // next we create a `UserDetails` object for each record. this involves looking up
+        // the user in the delius oracle db in order to get the user id. users which exist
+        // in the LDAP but not in the delius oracle db are not included in the result.
+        return userList.stream()
+            .map(user -> {
+                uk.gov.justice.digital.delius.jpa.national.entity.User oracleUser;
+                final String userCn = user.getCn();
+
+                try {
+                    oracleUser = userRepositoryWrapper.getUser(userCn);
+                } catch (NoSuchUserException e) {
+                    log.error("no entry found in delius USER table for LDAP user '{}'", userCn);
+                    return null;
+                }
+
+                return UserDetails.builder()
+                    .roles(user.getRoles().stream().map(role -> UserRole.builder().name(role.getCn()).build()).collect(toList()))
+                    .firstName(user.getGivenname())
+                    .surname(user.getSn())
+                    .email(user.getMail())
+                    .enabled(user.isEnabled())
+                    .userId(oracleUser.getUserId())
+                    .build();
+            })
+            .filter(user -> user != null)
+            .collect(Collectors.toList());
```

Comment:
I wonder if this would read slightly better if you extract a function?



```        return userList.stream()
            .map(user -> UserDetails.builder()
                    .roles(user.getRoles().stream().map(role -> UserRole.builder().name(role.getCn()).build()).collect(toList()))
                    .firstName(user.getGivenname())
                    .surname(user.getSn())
                    .email(user.getMail())
                    .enabled(user.isEnabled())
                    .userId(userIdOf(user.getCn()))
                    .build();
            )
            .filter(user -> user.getUserId() != null)
            .collect(Collectors.toList());`

private Long userIdOf(String cn) {
                try {
                    return userRepositoryWrapper.getUser(userCn).getUserId();
                } catch (NoSuchUserException e) {
                    log.error("no entry found in delius USER table for LDAP user '{}'", userCn);
                    return null;
                }

}
```

Reply:
like the idea, but i would want to reuse the `userIdOf` method in `getUserDetails` too, but in that case we don't want to swallow the exception. this makes me think the implementation is specific enough to just stay where it is.